nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-client-service/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchLookupService.java

```diff
@@ -0,0 +1,290 @@
+        List<String> reasons = new ArrayList<>();
+
+        if (coordinates.containsKey("_id") && !(coordinates.get("_id") instanceof String)) {
+            reasons.add("_id was supplied, but it was not a String.");
+        } else if (coordinates.containsKey("query") && !(coordinates.get("query") instanceof String)) {
+            reasons.add("query was supplied, but it was not a String.");
+        } else if (!coordinates.containsKey("_id") && !coordinates.containsKey("query")) {
+            reasons.add("Either \"_id\" or \"query\" must be supplied as keys to lookup(Map)");
+        } else if (coordinates.containsKey("_id") && coordinates.containsKey("query")) {
+            reasons.add("\"_id\" and \"query\" cannot be used at the same time as keys.");
+        }
+
+        if (reasons.size() > 0) {
+            String error = String.join("\n", reasons);
+            throw new LookupFailureException(error);
+        }
+    }
+
+    private Record getById(final String _id) throws IOException, LookupFailureException {
+        Map<String, Object> query = new HashMap<String, Object>(){{
+            put("query", new HashMap<String, Object>() {{
+                put("match", new HashMap<String, String>(){{
+                    put("_id", _id);
+                }});
+            }});
+        }};
+
+        String json = mapper.writeValueAsString(query);
+
+        SearchResponse response = clientService.search(json, index, type);
+
+        if (response.getNumberOfHits() > 1) {
+            throw new LookupFailureException(String.format("Expected 1 response, got %d for query %s",
+                response.getNumberOfHits(), json));
+        } else if (response.getNumberOfHits() == 0) {
+            return null;
+        }
+
+        final Map<String, Object> source = (Map)response.getHits().get(0).get("_source");
+
+        RecordSchema toUse = recordSchema != null ? recordSchema : convertSchema(source);
+
+        return new MapRecord(toUse, source);
+    }
+
+    private Record getByQuery(final String query) throws LookupFailureException {
+        Map<String, Object> parsed;
+        try {
+            parsed = mapper.readValue(query, Map.class);
+            parsed.remove("from");
+            parsed.remove("aggs");
+            parsed.put("size", 1);
+
+            final String json = mapper.writeValueAsString(parsed);
+
+            SearchResponse response = clientService.search(json, index, type);
+
+            if (response.getNumberOfHits() == 0) {
+                return null;
+            } else {
+                final Map<String, Object> source = (Map)response.getHits().get(0).get("_source");
+                RecordSchema toUse = recordSchema != null ? recordSchema : convertSchema(source);
+                return new MapRecord(toUse, source);
+            }
+
+        } catch (IOException e) {
+            throw new LookupFailureException(e);
+        }
+    }
+
+    private RecordSchema convertSchema(Map<String, Object> result) {
```

Comment:
Should this be in a utilities class (if there isn't already such a method in one)? Seems pretty helpful for JSON-to-schema conversions (or any Map-to-Schema) in general.

Reply:
I agree. I just wrote up [a Jira ticket](https://issues.apache.org/jira/browse/NIFI-5127) for this. Let's table it for now because we need to think about things like date strings.